### PR TITLE
Project HPS, first pass

### PR DESCRIPTION
**Pull request by saghul**

### CODE REVIEW

File: JitsiConference.js [hunk 1]
```diff
@@ -1636,6 +1653,18 @@ JitsiConference.prototype.onRemoteTrackAdded = function(track) {
         return;
     }
 
+    // Setup E2EE handling, if supported.
+    if (this._e2eeCtx) {
+        const activeTPC = this.getActivePeerConnection();
```
Comment (paweldomas) [inline JitsiConference.js:1658]: is E2EE supposed to be used only with the JVB connection?

Comment (saghul) [reply JitsiConference.js:1658]: Yes

File: JitsiConference.js [hunk 2]
```diff
@@ -233,6 +234,10 @@ export default function JitsiConference(options) {
     this.videoSIPGWHandler = new VideoSIPGW(this.room);
     this.recordingManager = new RecordingManager(this.room);
     this._conferenceJoinAnalyticsEventSent = false;
+
+    if (browser.supportsInsertableStreams()) {
```
Comment (fippo) [inline JitsiConference.js:238]: should there be an extra feature flag for this?

File: modules/e2ee/E2EEContext.js [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+     */
+    _encodeFunction(encodedFrame, controller) {
+        const keyIndex = this._currentKeyIndex % this._cryptoKeyRing.length;
+
+        if (this._cryptoKeyRing[keyIndex]) {
+            const iv = this._makeIV(encodedFrame.synchronizationSource, encodedFrame.timestamp);
+
+            return crypto.subtle.encrypt({
+                name: 'AES-GCM',
+                iv
```
Comment (fippo) [inline modules/e2ee/E2EEContext.js:234]: note: still need to pass additionalData here as suggested by @JonathanLennox

File: modules/e2ee/E2EEContext.js [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+                newUint8.set(new Uint8Array(encodedFrame.data, 0, unencryptedBytes[encodedFrameType]));
+                newUint8.set(new Uint8Array(plainText), unencryptedBytes[encodedFrameType]);
+
+                encodedFrame.data = newData;
+
+                return controller.enqueue(encodedFrame);
+            }, e => {
+                logger.error(e);
+
+                // Just feed the encoded frame in case of error.
```
Comment (fippo) [inline modules/e2ee/E2EEContext.js:316]: nit: "still encrypted" or "encrypted-but-not-decryptable garbage"

File: modules/e2ee/E2EEContext.js [hunk 5]
```diff
@@ -0,0 +1,325 @@
+/* global __filename, TransformStream */
+
+import { getLogger } from 'jitsi-meet-logger';
+
+const logger = getLogger(__filename);
+
+// We use a ringbuffer of keys so we can change them and still decode packets that were
+// encrypted with an old key.
+// In the future when we dont rely on a globally shared we will actually use it. For
```
Comment (fippo) [inline modules/e2ee/E2EEContext.js:9]: on a globally shared key. Silly me...

File: JitsiConference.js (hunk 2 above)
Comment (saghul) [reply JitsiConference.js:238]: No need for now.

**Pull request merged**